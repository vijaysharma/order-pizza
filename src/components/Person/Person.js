import React from 'react';
import CounterButton from '../../ui/CounterButton/CounterButton';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faChild, faUser} from '@fortawesome/free-solid-svg-icons';
import * as classes from './Person.module.css';

const Person = props => {
  const label = props.label ? props.label.toUpperCase() : '';
  return (
    <div className={classes.Person}>
      <span><FontAwesomeIcon icon={props.label === 'adults' ? faUser : faChild} /></span>
      <h2>{label}</h2>
      <div>
        <CounterButton
          count={props.count}
          increment={props.onIncrement}
          decrement={props.onDecrement}
          disabled={props.disabled} />
      </div>
    </div>
  );
};

export default Person;