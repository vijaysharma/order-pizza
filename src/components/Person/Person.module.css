.Person {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem 1rem 1rem;
  border-top: 2px solid #8b8b8b;
  margin-top: 1rem;
}
.Person:last-child {
  margin-top: 0;
  padding-bottom: 0;
}
.Person > span {
  font-size: 2rem;
  width: 3rem;
}

.Person > h2 {
  font-size: 1.5rem;
  font-weight: 400;
  margin: 0;
  flex-grow: 1;
}

.Person > div {
  text-align: right;
  flex-grow: 1;
}