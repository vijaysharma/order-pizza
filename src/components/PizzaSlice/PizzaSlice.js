import React from 'react';
import PizzaPortion from '../../ui/PizzaPortion/PizzaPortion';
import CounterButton from '../../ui/CounterButton/CounterButton';
import * as classes from './PizzaSlice.module.css';

const PizzaSlice = props => {
  return (
    <div className={classes.PizzaSlice}>
      <PizzaPortion
        size={props.size}
        label={props.label} />
      <CounterButton
        count={props.count}
        increment={props.onIncrement}
        decrement={props.onDecrement}
        disabled={props.disabled} />
    </div>
  );
};

export default PizzaSlice;