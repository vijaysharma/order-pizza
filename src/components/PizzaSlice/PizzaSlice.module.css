.PizzaSlice {
  text-align: right;
  padding: 1rem;
}
@media only screen and (max-width: 640px) {
  .PizzaSlice {
    padding: .75rem;
  }
}