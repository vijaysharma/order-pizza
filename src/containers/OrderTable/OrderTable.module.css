.OrderTableWrapper {
  color: #283688;
  width: 60%;
  margin: 3rem auto;
}

.OrderTable {
  border: 2px solid #8b8b8b;
  padding: 1rem;
}

.OrderTitle {
  font-size: 1.5rem;
  font-weight: 300;
}

.OrderTitle span {
  font-weight: 700;
}
.OrderTotal {
  display: flex;
  align-content: center;
  padding: 1.5rem;
}
.OrderTotal > h4,
.OrderTotal > span {
  flex-grow: 1;
}
.OrderTotal > h4 {
  margin: 0;
}
.OrderTotal > span {
  color: #444;
  line-height: 2rem;
  text-align: right;
  padding-right: 3.5rem;
}
@media only screen and (max-width: 640px) {
  .OrderTableWrapper {
    width: 100%;
    padding: 1rem;
  }
  .OrderTable {
    padding: 0;
  }
}