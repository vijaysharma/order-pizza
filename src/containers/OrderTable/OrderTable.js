import React, {useState} from 'react';
import PizzaSlice from '../../components/PizzaSlice/PizzaSlice';
import Person from '../../components/Person/Person';
import * as classes from './OrderTable.module.css';

const OrderTable = () => {
  const PRICE = {
    small: 150,
    medium: 200,
    large: 300
  };
  const [pizzaState, setPizzaState] = useState({small: 0, medium: 1, large: 0});
  const [personState, setPersonState] = useState({adults: 1, children: 0});
  const [totalPrice, setTotalPrice] = useState(200);

  const calculateTotalPrice = (state, price) => {
    const totalPrice = Object.keys(state).reduce((total, key) => {
      return total + state[key] * price[key];
    }, 0);
    return totalPrice;
  };
  const getTotalSmall = (state) => {
    const stateCopy = {...state};
    return stateCopy.large * 4 + stateCopy.medium * 2 + stateCopy.small;
  };

  const pizzaBoundary = (newState) => {
    let newStateCopy = {...newState};
    const newTotalSmalls = getTotalSmall(newStateCopy);
    if (newTotalSmalls < 2) {
      newStateCopy.medium = 1;
    }
    return newStateCopy;
  };

  const pizzaBestCombination = (state) => {
    const totalSmalls = getTotalSmall(state);
    const large = Math.floor(totalSmalls / 4);
    const medium = Math.floor((totalSmalls - large * 4) / 2);
    const small = (totalSmalls - large * 4 - medium * 2);

    return {small, medium, large};
  };

  const changePersonStateOnPizza = (pizzas, persons) => {
    const copyPizzas = {...pizzas};
    const copyPersons = {...persons};
    let totalPossibleSmalls = getTotalSmall(copyPizzas);
    const totalPossibleChildren = copyPersons.adults * 2 + copyPersons.children;

    if (totalPossibleSmalls < totalPossibleChildren) {
      if (copyPersons.adults >= 1) {
        copyPersons.adults = Math.floor((totalPossibleSmalls - copyPersons.children) / 2) <= 0 ? 1 : Math.floor((totalPossibleSmalls - copyPersons.children) / 2);
        copyPersons.children = (totalPossibleChildren - copyPersons.adults * 2) > (totalPossibleSmalls - copyPersons.adults * 2) ? (totalPossibleSmalls - copyPersons.adults * 2) : (totalPossibleChildren - copyPersons.adults * 2);
      } else {
        copyPersons.adults = 1;
        copyPersons.children = totalPossibleChildren - 2;
      }
    }
    return copyPersons;
  };

  const incrementPizzaHandler = (label) => {
    let newPizzaState = {...pizzaState};
    newPizzaState[label] = newPizzaState[label] + 1;
    newPizzaState = pizzaBestCombination(newPizzaState);
    setPizzaState(newPizzaState);

    const newPrice = calculateTotalPrice(newPizzaState, PRICE);
    setTotalPrice(newPrice);
  };

  const decrementPizzaHandler = (label) => {
    if (pizzaState[label] === 0) {
      return;
    }
    let newPizzaState = {...pizzaState};
    newPizzaState[label] = newPizzaState[label] - 1;
    if (newPizzaState[label] <= 0) {
      newPizzaState[label] = 0;
    }
    newPizzaState = pizzaBoundary(newPizzaState);
    newPizzaState = pizzaBestCombination(newPizzaState);
    setPizzaState(newPizzaState);

    const newPersonState = changePersonStateOnPizza(newPizzaState, personState);
    setPersonState(newPersonState);

    const newPrice = calculateTotalPrice(newPizzaState, PRICE);
    setTotalPrice(newPrice);
  };

  const personBoundary = (newState) => {
    const newStateCopy = {...newState};
    const totalAdults = newStateCopy.adults;
    if (totalAdults < 1) {
      newStateCopy.adults = 1;
    }
    return newStateCopy;
  };

  const changePizzaStateOnPerson = (pizzas, persons) => {
    let copyPizzas = {...pizzas};
    const copyPersons = {...persons};
    const totalPossibleSmalls = getTotalSmall(pizzas);
    const totalPossibleChildren = copyPersons.adults * 2 + copyPersons.children;

    if (totalPossibleChildren > totalPossibleSmalls) {
      copyPizzas = pizzaBestCombination({small: totalPossibleChildren, medium: 0, large: 0});
    }
    return copyPizzas;
  };

  const incrementPersonHandler = (label) => {
    let newPersonState = {...personState};
    newPersonState[label] = newPersonState[label] + 1;
    setPersonState(newPersonState);

    const newPizzaState = changePizzaStateOnPerson(pizzaState, newPersonState);
    setPizzaState(newPizzaState);

    const newPrice = calculateTotalPrice(newPizzaState, PRICE);
    setTotalPrice(newPrice);
  };

  const decrementPersonHandler = (label) => {
    if (personState[label] === 0) {
      return;
    }
    let newPersonState = {...personState};
    newPersonState[label] = newPersonState[label] - 1;
    if (newPersonState[label] <= 0) {
      newPersonState[label] = 0;
    }
    newPersonState = personBoundary(newPersonState);
    setPersonState(newPersonState);

    let newPizzaState = changePizzaStateOnPerson(pizzaState, newPersonState);
    setPizzaState(newPizzaState);

    const newPrice = calculateTotalPrice(newPizzaState, PRICE);
    setTotalPrice(newPrice);
  };


  const PizzaSlices = Object.keys(pizzaState).map(pizzaSize => (
      <PizzaSlice
        key={pizzaSize}
        size={pizzaSize}
        label={pizzaSize}
        count={pizzaState[pizzaSize]}
        onIncrement={() => incrementPizzaHandler(pizzaSize)}
        onDecrement={() => decrementPizzaHandler(pizzaSize)}
        disabled={!pizzaState[pizzaSize]} />
    )
  );

  const Persons = Object.keys(personState).map(person => (
      <Person
        key={person}
        label={person}
        count={personState[person]}
        onIncrement={() => incrementPersonHandler(person)}
        onDecrement={() => decrementPersonHandler(person)}
        disabled={!personState[person]} />
    )
  );

  return (
    <div className={classes.OrderTableWrapper}>
      <h1 className={classes.OrderTitle}>Order <span>Pizza</span></h1>
      <div className={classes.OrderTable}>
        {PizzaSlices}
        {Persons}
      </div>
      <div className={classes.OrderTotal}>
        <h4 className={classes.OrderTitle}>Order <span>Total</span></h4><span>{totalPrice}</span>
      </div>
    </div>
  );
};

export default OrderTable;