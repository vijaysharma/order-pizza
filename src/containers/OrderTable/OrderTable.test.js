import React from 'react';

import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import OrderTable from './OrderTable';
import PizzaSlice from '../../components/PizzaSlice/PizzaSlice';
import Person from '../../components/Person/Person';
import * as ButtonClasses from '../../ui/CounterButton/CounterButton.module.css';

configure({adapter: new Adapter()});

describe('<OrderTable/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<OrderTable />);
  });

  it('should render <OrderTable/> component', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should render <OrderTable/> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render <OrderTable/> component with 3 PizzaSlice', () => {
    expect(wrapper.find(PizzaSlice)).toHaveLength(3);
  });

  it('should render <OrderTable/> component with 2 Person', () => {
    expect(wrapper.find(Person)).toHaveLength(2);
  });

});