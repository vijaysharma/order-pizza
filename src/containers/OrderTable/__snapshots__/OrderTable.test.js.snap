// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderTable/> should render <OrderTable/> correctly 1`] = `ShallowWrapper {}`;
